src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -4135,6 +4160,26 @@ WorkerLimitCount(MultiExtendedOp *originalOpNode)
 								workerLimitCountInt64)));
 	}
 
+	/* if offset is present, add offset value to limit on workers */
+	if (originalOpNode->limitOffset != NULL)
+	{
+		Const *workerOffsetConst = (Const *) originalOpNode->limitOffset;
+		Const *workerLimitConst = NULL;
+		int64 workerOffsetValueInt64 = DatumGetInt64(workerOffsetConst->constvalue);
+		int64 workerNewLimitValueInt64 = -1;
+
+		/* if a limit is not present, create a new node for limit */
+		if (workerLimitCount == NULL)
+		{
+			workerLimitCount = (Node *) MakeIntegerConstInt64(0);
+		}
+
+		workerLimitConst = (Const *) workerLimitCount;
+		workerNewLimitValueInt64 = DatumGetInt64(workerLimitConst->constvalue);
+		workerNewLimitValueInt64 += workerOffsetValueInt64;
+		workerLimitConst->constvalue = Int64GetDatum(workerNewLimitValueInt64);
+	}
+
```

Comment:
Would something like this be simpler? We always sum limit and offset, and if the limit is null, use the initial value 0. 


```sql
	if (originalOpNode->limitOffset != NULL)
	{
		Const *workerOffsetConst = (Const *) originalOpNode->limitOffset;
		int64 workerOffsetValueInt64 = DatumGetInt64(workerOffsetConst->constvalue);
		int64 workerLimitValueInt64 = 0;

		/* if a limit is present, use it for initial value */
		if (workerLimitCount != NULL)
		{
			Const *workerLimitConst = (Const *) workerLimitCount;
			workerLimitValueInt64 = DatumGetInt64(workerLimitConst->constvalue);
		}

		workerLimitValueInt64 = workerLimitValueInt64 + workerOffsetValueInt64;
		workerLimitCount = (Node *) MakeIntegerConstInt64(workerLimitValueInt64);
	}
```